helix-gateway/src/main/java/org/apache/helix/gateway/util/GatewayCurrentStateCache.java

```diff
@@ -47,111 +51,113 @@ public GatewayCurrentStateCache(String clusterName) {
+      ShardStateMap oldStateMap = _currentStateMap.get(instance);
+      Map<String, Map<String, String>> instanceDiff = oldStateMap == null ? userCurrentStateMap.get(instance)
+          : oldStateMap.getDiff(userCurrentStateMap.get(instance));
+      if (!instanceDiff.isEmpty()) {
+        diff.put(instance, instanceDiff);
       }
+      newCurrentStateMap.put(instance, new ShardStateMap(userCurrentStateMap.get(instance)));
     }
+    logger.info("Update current state cache for instances: {}", diff.keySet());
+    _currentStateMap = newCurrentStateMap;
     return diff;
   }
 
   /**
-   * Update the cache with the current state diff.
-   * All existing target states remains the same
-   * @param currentStateDiff
+   * Update the current state with the changed current state maps.
    */
-  public void updateCacheWithCurrentStateDiff(Map<String, Map<String, Map<String, String>>> currentStateDiff) {
-    for (String instance : currentStateDiff.keySet()) {
-      Map<String, Map<String, String>> currentStateDiffMap = currentStateDiff.get(instance);
-      updateShardStateMapWithDiff(_currentStateMap, instance, currentStateDiffMap);
-    }
+  public void updateCurrentStateOfExistingInstance(String instance, String resource, String shard, String shardState) {
+    logger.info("Update current state of instance: {}, resource: {}, shard: {}, state: {}", instance, resource, shard,
+        shardState);
+    updateShardStateMapWithDiff(_currentStateMap, instance, resource, shard, shardState);
   }
 
   /**
    * Update the target state with the changed target state maps.
    * All existing target states remains the same
-   * @param targetStateChangeMap
    */
-  public void updateTargetStateWithDiff(String instance, Map<String, Map<String, String>> targetStateChangeMap) {
-    updateShardStateMapWithDiff(_targetStateMap, instance, targetStateChangeMap);
+  public void updateTargetStateOfExistingInstance(String instance, String resource, String shard, String shardState) {
+    logger.info("Update target state of instance: {}, resource: {}, shard: {}, state: {}", instance, resource, shard,
+        shardState);
+    updateShardStateMapWithDiff(_targetStateMap, instance, resource, shard, shardState);
+  }
+
+  private void updateShardStateMapWithDiff(Map<String, ShardStateMap> stateMap, String instance, String resource,
+      String shard, String shardState) {
+    ShardStateMap curStateMap = stateMap.get(instance);
+    if (curStateMap == null) {
+      logger.warn("Instance {} is not in the state map, skip updating", instance);
+      return;
+    }
+    curStateMap.updateWithShardState(resource, shard, shardState);
   }
 
   /**
    * Serialize the target state assignments to a JSON Node.
    * example ： {"instance1":{"resource1":{"shard1":"ONLINE","shard2":"OFFLINE"}}}}
    */
-  public ObjectNode serializeTargetAssignmentsToJSONNode() {
+  public synchronized ObjectNode serializeTargetAssignmentsToJSONNode() {
```

Comment:
Optimize this? Create a snapshot object for get operation? And serailize data and update it when it is needed? I am afraid of having many expired objects increasing GC.

Reply:
Then won't create snapshot bringing a lot object for GC?